pkg/api/norman/customization/cred/store.go

```diff
@@ -107,3 +138,97 @@ func (s *Store) Delete(apiContext *types.APIContext, schema *types.Schema, id st
 	}
 	return s.Store.Delete(apiContext, schema, id)
 }
+
+// processHarvesterCloudCredentialExpiration will extract the kubeconfig from a harvester cloud credential, extract the
+// expiration in milliseconds from the associated token, and set the CloudCredentialExpirationAnnotation on the
+// secret, removing on update if not present. Cloud credentials are transformed by norman, and thus the secret key of
+// `harvestercredentialConfig-kubeconfigContent` is unrolled into a top level map with `kubeconfigContent` as a nested
+// element within that map (see: decodeNonPasswordFields). If the credential is not a harvester credential, this
+// function does nothing.
+func (s *Store) processHarvesterCloudCredentialExpiration(data map[string]any, remove bool) error {
+	if hc, ok := data["harvestercredentialConfig"].(map[string]any); ok && hc != nil {
+		if kubeconfigYaml, ok := hc["kubeconfigContent"].(string); ok && kubeconfigYaml != "" {
+
+			expiration, err := GetHarvesterCloudCredentialExpirationFromKubeconfig(kubeconfigYaml, func(tokenName string) (*apimgmtv3.Token, error) {
+				return s.TokenLister.Get("", tokenName)
+			})
+			if err != nil {
+				return fmt.Errorf("failed to get harvester cloud credential expiration from kubeconfig: %w", err)
+			}
+
+			if expiration != "" {
+				values.PutValue(data, expiration, "annotations", CloudCredentialExpirationAnnotation)
+			} else if remove {
+				values.RemoveValue(data, "annotations", CloudCredentialExpirationAnnotation)
+			}
```

Comment:
Does this logic not handle the case where we do want to remove the annotation (passing remove as true) but the expiration is not an empty string?  Is that a valid case?

Reply:
Well if the annotation isn't empty, we don't want to remove it, we only want to remove it if it's empty and an update. I would argue that the comment does a pretty good job of explaining the behavior.